src/py21cmfast/wrapper.py

```diff
@@ -3219,3 +3229,248 @@ def calibrate_photon_cons(
+        if lc.velocity_x is None or lc.velocity_y is None:
+            logger.error(
+                "velocity_x and velocity_y are required! rerun the lightcone with OUTPUT_ALL_VEL=True"
+            )
+    else:
+        if user_params is None:
+            user_params = UserParams(**UserParams._defaults_)
+        if cosmo_params is None:
+            cosmo_params = CosmoParams(**CosmoParams._defaults_)
+        if astro_params is None:
+            astro_params = AstroParams(**AstroParams._defaults_)
+        if flag_options is None:
+            flag_options = FlagOptions(**FlagOptions._defaults_)
+        lc_quantities = (
+            "brightness_temp",
+            "xH_box",
+            "density",
+            "velocity",
+            "velocity_y",
+            "velocity_x",
+        )
+        lc = run_lightcone(
+            redshift=z_start,
+            user_params=user_params,
+            cosmo_params=cosmo_params,
+            astro_params=astro_params,
+            flag_options=flag_options,
+            lightcone_quantities=lc_quantities,
+            random_seed=random_seed,
+        )
+        logger.warning(
+            "run_kSZ requires a lightcone object, which is not given. Running with default parameters."
+        )
+    random.seed(random_seed)
+
+    kSZ_consts = _KszConstants(
+        lc.user_params.HII_DIM,
+        lc.user_params.BOX_LEN,
+        lc.cosmo_params.hlittle,
+        lc.cosmo_params.OMb,
+        len(lc.lightcone_redshifts),
+        z_start,
+        lc.lightcone_distances[0],
+        0.245,  # Helium fraction
+        3,  # redshift of helium reionization, for tau_e calculation
+    )
+
+    kSZ_consts.mean_taue_curr_z = compute_tau(
+        redshifts=[z_start],
+        global_xHI=[1],
+        user_params=user_params,
+        cosmo_params=cosmo_params,
+    )
+    Tcmb, mean_taue_fin = _Proj_array(
+        lc.lightcone_redshifts,
+        lc.density,
+        lc.velocity,
+        lc.xH_box,
+        kSZ_consts,
+        PARALLEL_APPROX=PARALLEL_APPROX,
+        rotation=rotation,
+    )
+
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore")
+        P_k, l_s, err = get_power(
+            Tcmb
+            * kSZ_consts.CMperMPC
+            / constants.c.cgs.value
+            * 1e6
+            * Planck15.Tcmb0.value
+            / np.sqrt(2 * np.pi),
+            lc.user_params.BOX_LEN,
+            bins=30,
+            log_bins=True,
+            get_variance=True,
+        )  # in microK^2
+    P_k = P_k * l_s ** 2
+    err = np.sqrt(err) * l_s ** 2
+    l_s *= lc.lightcone_distances[0]
+    return KSZOutput(
+        Tcmb * kSZ_consts.CMperMPC / constants.c.cgs.value * Planck15.Tcmb0.value,
+        mean_taue_fin,
+        l_s=l_s[np.logical_not(np.isnan(l_s))],
+        kSZ_power=P_k[np.logical_not(np.isnan(l_s))],
+        err=err[np.logical_not(np.isnan(l_s))],
+    )
+
+
+def _Proj_array(
+    redshifts,
+    density,
+    velocity,
```

Comment:
couldnt remember what we decided to do here. but I think the velocity here should be some kind of combination of velocity_x, velocity_y and velocity? or you need to pass all of them into this function and select randomly when you rotate the box?

Reply:
Well, if I'm rotating the lightcone, velocity will have all of the components incorporated, I will add that as well to run_lightcone. If I'm not rotating the lightcone, I'm still doing kSZ calculation over the z-component since that is the axis over which we're interpolating. So keeping only velocity should be fine.

Reply:
yeah, I'm just thinking if you do in in the LC (i.e. replace some part of vz with vx and vy when generate lc.velocity), would it complicate things when people also need to use vz for other things? (I can't think of any at the moment though...) But I know the benefit of doing it in the LC instead of in this function is that you will save RAM of two LC arrays (lc.velocity_x and _y)... 

Why am I having this dejavu that we have talked about this before? Have we had reached a conclusion already?